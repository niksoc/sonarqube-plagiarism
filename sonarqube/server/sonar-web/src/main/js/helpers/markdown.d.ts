/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
interface FrontMatter {
  [x: string]: string;
}

interface ParsedContent {
  content: string;
  frontmatter: FrontMatter;
}

export function getFrontMatter(content: string): FrontMatter;

export function separateFrontMatter(content: string): ParsedContent;

/** Removes SonarQube/SonarCloud only content */
export function filterContent(content: string): string;
