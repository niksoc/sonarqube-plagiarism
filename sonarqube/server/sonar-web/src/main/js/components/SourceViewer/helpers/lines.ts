/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { intersection } from 'lodash';

export function optimizeHighlightedSymbols(
  symbolsForLine: string[] = [],
  highlightedSymbols: string[] = []
): string[] | undefined {
  const symbols = intersection(symbolsForLine, highlightedSymbols);

  return symbols.length ? symbols : undefined;
}

export function optimizeLocationMessage(
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined,
  optimizedSecondaryIssueLocations: T.LinearIssueLocation[]
) {
  return highlightedLocationMessage != null &&
    optimizedSecondaryIssueLocations.some(
      location => location.index === highlightedLocationMessage.index
    )
    ? highlightedLocationMessage
    : undefined;
}

export function optimizeSelectedIssue(selectedIssue: string | undefined, issuesForLine: T.Issue[]) {
  return selectedIssue !== undefined && issuesForLine.find(issue => issue.key === selectedIssue)
    ? selectedIssue
    : undefined;
}
