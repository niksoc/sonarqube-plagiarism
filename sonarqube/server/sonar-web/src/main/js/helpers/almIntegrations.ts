/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';

export function getAlmMembersUrl(key: string, url: string): string {
  if (!url.endsWith('/')) {
    url += '/';
  }
  if (isGithub(key)) {
    return url.replace('github.com/', 'github.com/orgs/') + 'people';
  }
  return url + 'profile/members';
}

export function getUserAlmKey(user: T.CurrentUser) {
  return isLoggedIn(user) && user.externalProvider
    ? sanitizeAlmId(user.externalProvider)
    : undefined;
}

export function hasAdvancedALMIntegration(user: T.CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string): boolean {
  return almKey ? almKey.startsWith('bitbucket') : false;
}

export function isGithub(almKey?: string): boolean {
  return almKey === 'github';
}

export function isVSTS(almKey?: string): boolean {
  return almKey === 'microsoft';
}

export function sanitizeAlmId(almKey: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
