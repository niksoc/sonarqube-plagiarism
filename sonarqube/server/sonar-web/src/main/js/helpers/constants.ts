/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { colors } from '../app/theme';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT'
];
export const RULE_TYPES: T.RuleType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL', 'SECURITY_HOTSPOT'];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  colors.red,
  colors.orange,
  colors.yellow,
  colors.lightGreen,
  colors.green
];

export const RATING_COLORS = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];
