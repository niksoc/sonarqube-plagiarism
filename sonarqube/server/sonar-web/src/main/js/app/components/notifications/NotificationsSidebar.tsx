/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as differenceInSeconds from 'date-fns/difference_in_seconds';
import * as React from 'react';
import { ClearButton } from 'sonar-ui-common/components/controls/buttons';
import Modal from 'sonar-ui-common/components/controls/Modal';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { PrismicFeatureNews } from '../../../api/news';
import DateFormatter from '../../../components/intl/DateFormatter';

export interface Props {
  fetchMoreFeatureNews: () => void;
  loading: boolean;
  loadingMore: boolean;
  news: PrismicFeatureNews[];
  onClose: () => void;
  notificationsLastReadDate?: Date;
  paging?: T.Paging;
}

export default function NotificationsSidebar(props: Props) {
  const { loading, loadingMore, news, notificationsLastReadDate, paging } = props;
  const header = translate('embed_docs.whats_new');
  return (
    <Modal contentLabel={header} onRequestClose={props.onClose}>
      <div className="notifications-sidebar">
        <div className="notifications-sidebar-top">
          <h3>{header}</h3>
          <ClearButton
            className="button-tiny"
            iconProps={{ size: 12, thin: true }}
            onClick={props.onClose}
          />
        </div>
        <div className="notifications-sidebar-content">
          {loading ? (
            <div className="text-center">
              <DeferredSpinner className="big-spacer-top" timeout={200} />
            </div>
          ) : (
            news.map((slice, index) => (
              <Notification
                key={slice.publicationDate}
                notification={slice}
                unread={isUnread(index, slice.publicationDate, notificationsLastReadDate)}
              />
            ))
          )}
        </div>
        {!loading && paging && paging.total > news.length && (
          <div className="notifications-sidebar-footer">
            <div className="spacer-top note text-center">
              <a className="spacer-left" href="#" onClick={props.fetchMoreFeatureNews}>
                {translate('show_more')}
              </a>
              {loadingMore && (
                <DeferredSpinner className="text-bottom spacer-left position-absolute" />
              )}
            </div>
          </div>
        )}
      </div>
    </Modal>
  );
}

export function isUnread(index: number, notificationDate: string, lastReadDate?: Date) {
  return !lastReadDate ? index < 1 : differenceInSeconds(notificationDate, lastReadDate) > 0;
}

interface NotificationProps {
  notification: PrismicFeatureNews;
  unread: boolean;
}

export function Notification({ notification, unread }: NotificationProps) {
  return (
    <div className={classNames('notifications-sidebar-slice', { unread })}>
      <h4>
        <DateFormatter date={notification.publicationDate} long={false} />
      </h4>
      {notification.features.map((feature, index) => (
        <Feature feature={feature} key={index} />
      ))}
    </div>
  );
}

interface FeatureProps {
  feature: PrismicFeatureNews['features'][0];
}

export function Feature({ feature }: FeatureProps) {
  return (
    <div className="feature">
      <ul className="categories spacer-bottom">
        {feature.categories.map(category => (
          <li key={category.name} style={{ backgroundColor: category.color }}>
            {category.name}
          </li>
        ))}
      </ul>
      <span>{feature.description}</span>
      {feature.readMore && (
        <a
          className="learn-more"
          href={feature.readMore}
          rel="noopener noreferrer nofollow"
          target="_blank">
          {translate('learn_more')}
        </a>
      )}
    </div>
  );
}
