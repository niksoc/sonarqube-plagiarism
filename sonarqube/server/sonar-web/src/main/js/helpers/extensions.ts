/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { getExtensionFromCache } from './extensionsHandler';

let librariesExposed = false;

export function installScript(url: string, target: 'body' | 'head' = 'body'): Promise<any> {
  return new Promise(resolve => {
    const scriptTag = document.createElement('script');
    scriptTag.src = `${getBaseUrl()}${url}`;
    scriptTag.onload = resolve;
    document.getElementsByTagName(target)[0].appendChild(scriptTag);
  });
}

export async function getExtensionStart(key: string) {
  const fromCache = getExtensionFromCache(key);
  if (fromCache) {
    return Promise.resolve(fromCache);
  }

  if (!librariesExposed) {
    // Async import allows to reduce initial vendor bundle size
    const exposeLibraries = (await import('../app/components/extensions/exposeLibraries')).default;
    exposeLibraries();
    librariesExposed = true;
  }

  await installScript(`/static/${key}.js`);

  const start = getExtensionFromCache(key);
  if (start) {
    return start;
  }
  return Promise.reject();
}
