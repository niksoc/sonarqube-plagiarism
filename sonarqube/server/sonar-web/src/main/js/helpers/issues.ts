/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { flatten, sortBy } from 'lodash';
import { ISSUE_TYPES } from './constants';

interface Comment {
  login: string;
  [x: string]: any;
}

interface Rule {}

interface Component {
  key: string;
  name: string;
}

interface IssueBase {
  severity: string;
  [x: string]: any;
}

export interface RawIssue extends IssueBase {
  assignee?: string;
  author?: string;
  comments?: Array<Comment>;
  component: string;
  flows?: Array<{
    // `componentName` is not available in RawIssue
    locations?: Array<T.Omit<T.FlowLocation, 'componentName'>>;
  }>;
  key: string;
  line?: number;
  project: string;
  rule: string;
  status: string;
  subProject?: string;
  textRange?: T.TextRange;
}

export function sortByType<T extends Pick<T.Issue, 'type'>>(issues: T[]): T[] {
  return sortBy(issues, issue => ISSUE_TYPES.indexOf(issue.type));
}

function injectRelational(
  issue: T.Dict<any>,
  source: any[] | undefined,
  baseField: string,
  lookupField: string
) {
  const newFields: T.Dict<any> = {};
  const baseValue = issue[baseField];
  if (baseValue !== undefined && source !== undefined) {
    const lookupValue = source.find(candidate => candidate[lookupField] === baseValue);
    if (lookupValue != null) {
      Object.keys(lookupValue).forEach(key => {
        const newKey = baseField + key.charAt(0).toUpperCase() + key.slice(1);
        newFields[newKey] = lookupValue[key];
      });
    }
  }
  return newFields;
}

function injectCommentsRelational(issue: RawIssue, users?: T.UserBase[]) {
  if (!issue.comments) {
    return {};
  }
  const comments = issue.comments.map(comment => {
    const commentWithAuthor = { ...comment, author: comment.login, login: undefined };
    return {
      ...commentWithAuthor,
      ...injectRelational(commentWithAuthor, users, 'author', 'login')
    };
  });
  return { comments };
}

function prepareClosed(
  issue: RawIssue,
  secondaryLocations: T.FlowLocation[],
  flows: T.FlowLocation[][]
) {
  return issue.status === 'CLOSED'
    ? { flows: [], line: undefined, textRange: undefined, secondaryLocations: [] }
    : { flows, secondaryLocations };
}

function ensureTextRange(issue: RawIssue): { textRange?: T.TextRange } {
  return issue.line && !issue.textRange
    ? {
        textRange: {
          startLine: issue.line,
          endLine: issue.line,
          startOffset: 0,
          endOffset: 999999
        }
      }
    : {};
}

function reverseLocations(locations: T.FlowLocation[]): T.FlowLocation[] {
  const x = [...locations];
  x.reverse();
  return x;
}

function splitFlows(
  issue: RawIssue,
  components: Component[] = []
): { secondaryLocations: T.FlowLocation[]; flows: T.FlowLocation[][] } {
  const parsedFlows: T.FlowLocation[][] = (issue.flows || [])
    .filter(flow => flow.locations !== undefined)
    .map(flow => flow.locations!.filter(location => location.textRange != null))
    .map(flow =>
      flow.map(location => {
        const component = components.find(component => component.key === location.component);
        return { ...location, componentName: component && component.name };
      })
    );

  const onlySecondaryLocations = parsedFlows.every(flow => flow.length === 1);

  return onlySecondaryLocations
    ? { secondaryLocations: orderLocations(flatten(parsedFlows)), flows: [] }
    : { secondaryLocations: [], flows: parsedFlows.map(reverseLocations) };
}

function orderLocations(locations: T.FlowLocation[]) {
  return sortBy(
    locations,
    location => location.textRange && location.textRange.startLine,
    location => location.textRange && location.textRange.startOffset
  );
}

export function parseIssueFromResponse(
  issue: RawIssue,
  components?: Component[],
  users?: T.UserBase[],
  rules?: Rule[]
): T.Issue {
  const { secondaryLocations, flows } = splitFlows(issue, components);
  return {
    ...issue,
    ...injectRelational(issue, components, 'component', 'key'),
    ...injectRelational(issue, components, 'project', 'key'),
    ...injectRelational(issue, components, 'subProject', 'key'),
    ...injectRelational(issue, rules, 'rule', 'key'),
    ...injectRelational(issue, users, 'assignee', 'login'),
    ...injectCommentsRelational(issue, users),
    ...prepareClosed(issue, secondaryLocations, flows),
    ...ensureTextRange(issue)
  } as T.Issue;
}
