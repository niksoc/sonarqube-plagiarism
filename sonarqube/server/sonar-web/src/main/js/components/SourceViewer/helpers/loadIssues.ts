/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { searchIssues } from '../../../api/issues';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { parseIssueFromResponse } from '../../../helpers/issues';
import { BranchLike } from '../../../types/branch-like';

// maximum possible value
const PAGE_SIZE = 500;

function buildQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    additionalFields: '_all',
    resolved: 'false',
    componentKeys: component,
    s: 'FILE_LINE',
    ...getBranchLikeQuery(branchLike)
  };
}

export function loadPage(
  query: T.RawQuery,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize
  }).then(r =>
    r.issues.map(issue => parseIssueFromResponse(issue, r.components, r.users, r.rules))
  );
}

export function loadPageAndNext(
  query: T.RawQuery,
  toLine: number,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return loadPage(query, page).then(issues => {
    if (issues.length === 0) {
      return [];
    }

    const lastIssue = issues[issues.length - 1];

    if (
      (lastIssue.textRange != null && lastIssue.textRange.endLine > toLine) ||
      issues.length < pageSize
    ) {
      return issues;
    }

    return loadPageAndNext(query, toLine, page + 1, pageSize).then(nextIssues => {
      return [...issues, ...nextIssues];
    });
  });
}

export default function loadIssues(
  component: string,
  _fromLine: number,
  toLine: number,
  branchLike: BranchLike | undefined
): Promise<T.Issue[]> {
  const query = buildQuery(component, branchLike);
  return new Promise(resolve => {
    loadPageAndNext(query, toLine, 1).then(issues => {
      resolve(issues);
    });
  });
}
