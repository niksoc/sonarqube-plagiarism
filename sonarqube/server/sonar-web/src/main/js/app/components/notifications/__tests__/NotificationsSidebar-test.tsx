/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { parseDate } from 'sonar-ui-common/helpers/dates';
import NotificationsSidebar, {
  Feature,
  isUnread,
  Notification,
  Props
} from '../NotificationsSidebar';

const news: Props['news'] = [
  {
    notification: '10 Java rules, Github checks, Security Hotspots, BitBucket branch decoration',
    publicationDate: '2018-04-06',
    features: [
      {
        categories: [
          { color: '#ff0000', name: 'Java' },
          { color: '#00ff00', name: 'Rules' }
        ],
        description: '10 new Java rules'
      },
      {
        categories: [{ color: '#0000ff', name: 'BitBucket' }],
        description: 'BitBucket branch decoration',
        readMore: 'http://example.com'
      }
    ]
  },
  {
    notification: 'Some other notification',
    publicationDate: '2018-04-05',
    features: [
      {
        categories: [{ color: '#0000ff', name: 'BitBucket' }],
        description: 'BitBucket branch decoration',
        readMore: 'http://example.com'
      }
    ]
  }
];

describe('#NotificationSidebar', () => {
  it('should render correctly if there are new features', () => {
    const wrapper = shallowRender({ loading: true });
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ loading: false });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Notification')).toHaveLength(2);
  });

  it('should render correctly if there are no new unread features', () => {
    const wrapper = shallowRender({
      notificationsLastReadDate: parseDate('2018-12-31')
    });
    expect(wrapper.find('Notification')).toHaveLength(2);
    expect(wrapper.find('Notification[unread=true]')).toHaveLength(0);
  });
});

describe('#isUnread', () => {
  it('should be unread', () => {
    expect(isUnread(0, '2018-12-14', undefined)).toBe(true);
    expect(isUnread(1, '2018-12-14', parseDate('2018-12-12'))).toBe(true);
  });

  it('should be read', () => {
    expect(isUnread(0, '2018-12-16', parseDate('2018-12-16'))).toBe(false);
    expect(isUnread(1, '2018-12-15', undefined)).toBe(false);
  });
});

describe('#Notification', () => {
  it('should render correctly', () => {
    expect(shallow(<Notification notification={news[1]} unread={false} />)).toMatchSnapshot();
    expect(shallow(<Notification notification={news[1]} unread={true} />)).toMatchSnapshot();
  });
});

describe('#Feature', () => {
  it('should render correctly', () => {
    expect(shallow(<Feature feature={news[1].features[0]} />)).toMatchSnapshot();
    expect(shallow(<Feature feature={news[0].features[0]} />)).toMatchSnapshot();
  });
});

function shallowRender(props: Partial<Props> = {}) {
  return shallow(
    <NotificationsSidebar
      fetchMoreFeatureNews={jest.fn()}
      loading={false}
      loadingMore={false}
      news={news}
      notificationsLastReadDate={parseDate('2018-01-01')}
      onClose={jest.fn()}
      paging={{ pageIndex: 1, pageSize: 10, total: 20 }}
      {...props}
    />
  );
}
