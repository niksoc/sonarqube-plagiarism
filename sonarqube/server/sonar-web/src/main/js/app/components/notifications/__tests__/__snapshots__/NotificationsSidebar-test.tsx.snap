// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Feature should render correctly 1`] = `
<div
  className="feature"
>
  <ul
    className="categories spacer-bottom"
  >
    <li
      key="BitBucket"
      style={
        Object {
          "backgroundColor": "#0000ff",
        }
      }
    >
      BitBucket
    </li>
  </ul>
  <span>
    BitBucket branch decoration
  </span>
  <a
    className="learn-more"
    href="http://example.com"
    rel="noopener noreferrer nofollow"
    target="_blank"
  >
    learn_more
  </a>
</div>
`;

exports[`#Feature should render correctly 2`] = `
<div
  className="feature"
>
  <ul
    className="categories spacer-bottom"
  >
    <li
      key="Java"
      style={
        Object {
          "backgroundColor": "#ff0000",
        }
      }
    >
      Java
    </li>
    <li
      key="Rules"
      style={
        Object {
          "backgroundColor": "#00ff00",
        }
      }
    >
      Rules
    </li>
  </ul>
  <span>
    10 new Java rules
  </span>
</div>
`;

exports[`#Notification should render correctly 1`] = `
<div
  className="notifications-sidebar-slice"
>
  <h4>
    <DateFormatter
      date="2018-04-05"
      long={false}
    />
  </h4>
  <Feature
    feature={
      Object {
        "categories": Array [
          Object {
            "color": "#0000ff",
            "name": "BitBucket",
          },
        ],
        "description": "BitBucket branch decoration",
        "readMore": "http://example.com",
      }
    }
    key="0"
  />
</div>
`;

exports[`#Notification should render correctly 2`] = `
<div
  className="notifications-sidebar-slice unread"
>
  <h4>
    <DateFormatter
      date="2018-04-05"
      long={false}
    />
  </h4>
  <Feature
    feature={
      Object {
        "categories": Array [
          Object {
            "color": "#0000ff",
            "name": "BitBucket",
          },
        ],
        "description": "BitBucket branch decoration",
        "readMore": "http://example.com",
      }
    }
    key="0"
  />
</div>
`;

exports[`#NotificationSidebar should render correctly if there are new features 1`] = `
<Modal
  contentLabel="embed_docs.whats_new"
  onRequestClose={[MockFunction]}
>
  <div
    className="notifications-sidebar"
  >
    <div
      className="notifications-sidebar-top"
    >
      <h3>
        embed_docs.whats_new
      </h3>
      <ClearButton
        className="button-tiny"
        iconProps={
          Object {
            "size": 12,
            "thin": true,
          }
        }
        onClick={[MockFunction]}
      />
    </div>
    <div
      className="notifications-sidebar-content"
    >
      <div
        className="text-center"
      >
        <DeferredSpinner
          className="big-spacer-top"
          timeout={200}
        />
      </div>
    </div>
  </div>
</Modal>
`;

exports[`#NotificationSidebar should render correctly if there are new features 2`] = `
<Modal
  contentLabel="embed_docs.whats_new"
  onRequestClose={[MockFunction]}
>
  <div
    className="notifications-sidebar"
  >
    <div
      className="notifications-sidebar-top"
    >
      <h3>
        embed_docs.whats_new
      </h3>
      <ClearButton
        className="button-tiny"
        iconProps={
          Object {
            "size": 12,
            "thin": true,
          }
        }
        onClick={[MockFunction]}
      />
    </div>
    <div
      className="notifications-sidebar-content"
    >
      <Notification
        key="2018-04-06"
        notification={
          Object {
            "features": Array [
              Object {
                "categories": Array [
                  Object {
                    "color": "#ff0000",
                    "name": "Java",
                  },
                  Object {
                    "color": "#00ff00",
                    "name": "Rules",
                  },
                ],
                "description": "10 new Java rules",
              },
              Object {
                "categories": Array [
                  Object {
                    "color": "#0000ff",
                    "name": "BitBucket",
                  },
                ],
                "description": "BitBucket branch decoration",
                "readMore": "http://example.com",
              },
            ],
            "notification": "10 Java rules, Github checks, Security Hotspots, BitBucket branch decoration",
            "publicationDate": "2018-04-06",
          }
        }
        unread={true}
      />
      <Notification
        key="2018-04-05"
        notification={
          Object {
            "features": Array [
              Object {
                "categories": Array [
                  Object {
                    "color": "#0000ff",
                    "name": "BitBucket",
                  },
                ],
                "description": "BitBucket branch decoration",
                "readMore": "http://example.com",
              },
            ],
            "notification": "Some other notification",
            "publicationDate": "2018-04-05",
          }
        }
        unread={true}
      />
    </div>
    <div
      className="notifications-sidebar-footer"
    >
      <div
        className="spacer-top note text-center"
      >
        <a
          className="spacer-left"
          href="#"
          onClick={[MockFunction]}
        >
          show_more
        </a>
      </div>
    </div>
  </div>
</Modal>
`;
